source/sio/epoll/epoll_context.hpp

```diff
@@ -0,0 +1,897 @@
+      return fd;
+    }
+
+    struct context_base : stdexec::__immovable {
+      context_base()
+        : epoll_fd_(create_epoll())
+        , event_fd_(create_eventfd()) {
+        epoll_event event = {.events = EPOLLIN | EPOLLERR | EPOLLET, .data = {.ptr = &event_fd_}};
+        int ret = ::epoll_ctl(epoll_fd_, EPOLL_CTL_ADD, event_fd_, &event);
+        throw_error_code_if(ret < 0, errno);
+      }
+
+      ~context_base() {
+        epoll_event event = {};
+        int ret = ::epoll_ctl(epoll_fd_, EPOLL_CTL_DEL, event_fd_, &event);
+        throw_error_code_if(ret < 0, errno);
+      }
+
+      exec::safe_file_descriptor epoll_fd_;
+      exec::safe_file_descriptor event_fd_;
+    };
+
+    struct operation_base;
+
+    struct operation_vtable {
+      // return true means operation is done and associated context will complete operation.
+      bool (*execute_)(operation_base*) noexcept = nullptr;
+      void (*complete_)(operation_base*, const std::error_code&) noexcept = nullptr;
+    };
+
+    struct operation_base : stdexec::__immovable {
+      const operation_vtable* vtable_;
+      operation_base* next_{nullptr};
+
+      explicit operation_base(const operation_vtable& vtable)
+        : vtable_{&vtable} {
+      }
+    };
+
+    inline void stop_this_operation(operation_base* op) noexcept {
+      op->vtable_->complete_(op, std::make_error_code(std::errc::operation_canceled));
+    }
+
+    inline void complete_this_operation(operation_base* op) noexcept {
+      op->vtable_->complete_(op, std::error_code{});
+    }
+
+    inline bool execute_this_operation(operation_base* op) noexcept {
+      return op->vtable_->execute_(op);
+    }
+
+    using operation_queue = stdexec::__intrusive_queue<&operation_base::next_>;
+    using atomic_operation_queue = exec::__atomic_intrusive_queue<&operation_base::next_>;
+    using time_point = std::chrono::time_point<std::chrono::steady_clock>;
+
+    class scheduler;
+
+    class epoll_context : context_base {
+     public:
+      void run_until_stopped() {
+        // Only one thread of execution is allowed to drive the io context.
+        bool expected_running = false;
+
+        if (
+          !is_running_.compare_exchange_strong(expected_running, true, std::memory_order_relaxed)) {
+          throw std::runtime_error("sio::epoll::context::run() called on a running context");
+        } else {
+          // Check whether we restart the context after a context-wide stop.
+          // We have to reset the stop source in this case.
+          int in_flight = submissions_in_flight_.load(std::memory_order_relaxed);
+          if (in_flight == no_new_submissions) {
+            stop_source_.emplace();
+            // Make emplacement of stop source visible to other threads
+            // and open the door for new submissions.
+            submissions_in_flight_.store(0, std::memory_order_release);
+          }
+
+          // For eventfd.
+          epoll_submitted_ = 1;
```

Comment:
Hm in `io_uring` I have this in an else branch, I think. But thats a recent addition